dask/layers.py

```diff
@@ -154,17 +157,38 @@ def __iter__(self):
         return iter(self._dict)
 
     def __len__(self):
-        return len(self._dict)
+        interior_keys, _ = self._keys_from_overlap_operation()
+        n_tasks = 0
+        for k in interior_keys:
+            if all([isinstance(i, int) for i in k]):
+                n_tasks += 2  # represents both getitem and overlap tasks
+            else:
+                n_tasks += 1  # represents a single getitem task
+        return n_tasks
 
     def is_materialized(self):
         return hasattr(self, "_cached_dict")
 
     def get_output_keys(self):
-        return self.keys()  # FIXME! this implementation materializes the graph
+        if self._cached_output_keys is not None:
+            return self._cached_output_keys
 
-    def _dask_keys(self):
```

Comment:
I renamed the `_dask_keys` function to be `_keys_from_input_dependencies`, just because I thought that might be a little clearer (I was having a hard time keeping them straight in my head).

The git rendering of lines added/removed makes this seem more confusing than it is - if that's weird while reading maybe pull up the file here https://github.com/GenevieveBuckley/dask/blob/cccc80b633cb76639254ced7a96e1a67fa5adde6/dask/layers.py#L159-187